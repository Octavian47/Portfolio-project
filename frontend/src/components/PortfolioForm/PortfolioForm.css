/* src/components/PortfolioForm/PortfolioForm.css */

.message {
    padding: 10px;
    margin: 10px 0;
    border-radius: 5px;
    font-size: 16px;
  }
  
  .message.success {
    color: green;
    background-color: #e7f9e7;
  }
  
  .message.warning {
    color: orange;
    background-color: #fff4e5;
  }
  
  .message.error {
    color: red;
    background-color: #f9e7e7;
  }
  