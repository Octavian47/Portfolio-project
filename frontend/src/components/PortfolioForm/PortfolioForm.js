import React, { useState, useEffect } from 'react';
import './PortfolioForm.css';

const PortfolioForm = ({ initialData = {}, onSubmit }) => {
  const [title, setTitle] = useState(initialData.title || '');
  const [description, setDescription] = useState(initialData.description || '');
  const [clientUrl, setClientUrl] = useState(initialData.clientUrl || 'https://');
  const [status, setStatus] = useState(initialData.status || 'visible');
  const [imageFile, setImageFile] = useState(null);

  useEffect(() => {
    if (initialData.imageFilename) {
      setImageFile(initialData.imageFilename);
    }
  }, [initialData]);

  const handleSubmit = (e) => {
    e.preventDefault();
    onSubmit({ title, description, clientUrl, status, imageFile });
  };

  return (
    <form onSubmit={handleSubmit} className="portfolio-form">
      <div>
        <label htmlFor="title">Title:</label>
        <input
          id="title"
          type="text"
          value={title}
          onChange={(e) => setTitle(e.target.value)}
          required
        />
      </div>
      <div>
        <label htmlFor="description">Description:</label>
        <textarea
          id="description"
          value={description}
          onChange={(e) => setDescription(e.target.value)}
          required
        />
      </div>
      <div>
        <label htmlFor="clientUrl">Client URL:</label>
        <input
          id="clientUrl"
          type="url"
          value={clientUrl}
          onChange={(e) => setClientUrl(e.target.value)}
          placeholder="https://www.example.com"
          required
        />
        <small>Please include "https://" or "http://" in the URL.</small>
      </div>
      <div>
        <label htmlFor="status">Status:</label>
        <select
          id="status"
          value={status}
          onChange={(e) => setStatus(e.target.value)}
          required
        >
          <option value="visible">Display</option>
          <option value="hidden">Hidden</option>
        </select>
      </div>
      <div>
        <label htmlFor="image">Image:</label>
        <input
          id="image"
          type="file"
          onChange={(e) => setImageFile(e.target.files[0])}
        />
      </div>
      <button type="submit">Save</button>
    </form>
  );
};

export default PortfolioForm;
