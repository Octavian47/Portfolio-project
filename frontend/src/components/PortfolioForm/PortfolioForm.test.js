import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import PortfolioForm from './PortfolioForm';

describe('PortfolioForm Component', () => {

  const initialData = {
    title: 'Initial Title',
    description: 'Initial Description',
    clientUrl: 'https://www.example.com',
    status: 'visible',
    imageFilename: 'initialImage.jpg'
  };

  it('renders the form with all fields', () => {
    const { getByLabelText, getByText } = render(<PortfolioForm onSubmit={jest.fn()} />);

    expect(getByLabelText('Title:')).toBeInTheDocument();
    expect(getByLabelText('Description:')).toBeInTheDocument();
    expect(getByLabelText('Client URL:')).toBeInTheDocument();
    expect(getByLabelText('Status:')).toBeInTheDocument();
    expect(getByText('Save')).toBeInTheDocument();
  });

  it('updates the title field when typed into', () => {
    const { getByLabelText } = render(<PortfolioForm onSubmit={jest.fn()} />);
    
    const titleInput = getByLabelText('Title:');
    fireEvent.change(titleInput, { target: { value: 'New Title' } });
    
    expect(titleInput.value).toBe('New Title');
  });

  it('updates the description field when typed into', () => {
    const { getByLabelText } = render(<PortfolioForm onSubmit={jest.fn()} />);
    
    const descriptionInput = getByLabelText('Description:');
    fireEvent.change(descriptionInput, { target: { value: 'New Description' } });
    
    expect(descriptionInput.value).toBe('New Description');
  });

  it('updates the client URL field when typed into', () => {
    const { getByLabelText } = render(<PortfolioForm onSubmit={jest.fn()} />);
    
    const urlInput = getByLabelText('Client URL:');
    fireEvent.change(urlInput, { target: { value: 'https://www.new-url.com' } });
    
    expect(urlInput.value).toBe('https://www.new-url.com');
  });

  it('updates the status field when changed', () => {
    const { getByLabelText } = render(<PortfolioForm onSubmit={jest.fn()} />);
    
    const statusSelect = getByLabelText('Status:');
    fireEvent.change(statusSelect, { target: { value: 'hidden' } });
    
    expect(statusSelect.value).toBe('hidden');
  });

  it('handles file input correctly', () => {
    const { getByLabelText } = render(<PortfolioForm onSubmit={jest.fn()} />);
    
    const fileInput = getByLabelText('Image:');
    const file = new File(['dummy content'], 'example.png', { type: 'image/png' });
    
    fireEvent.change(fileInput, { target: { files: [file] } });
    
    expect(fileInput.files[0]).toBe(file);
  });

  it('calls onSubmit with correct data when form is submitted', () => {
    const handleSubmit = jest.fn();
    const { getByText, getByLabelText } = render(<PortfolioForm onSubmit={handleSubmit} />);
    
    fireEvent.change(getByLabelText('Title:'), { target: { value: 'New Title' } });
    fireEvent.change(getByLabelText('Description:'), { target: { value: 'New Description' } });
    fireEvent.change(getByLabelText('Client URL:'), { target: { value: 'https://www.new-url.com' } });
    fireEvent.change(getByLabelText('Status:'), { target: { value: 'hidden' } });

    const file = new File(['dummy content'], 'example.png', { type: 'image/png' });
    fireEvent.change(getByLabelText('Image:'), { target: { files: [file] } });

    fireEvent.click(getByText('Save'));

    expect(handleSubmit).toHaveBeenCalledWith({
      title: 'New Title',
      description: 'New Description',
      clientUrl: 'https://www.new-url.com',
      status: 'hidden',
      imageFile: file
    });
  });

  it('populates fields correctly when initialData is provided', () => {
    const { getByLabelText } = render(<PortfolioForm initialData={initialData} onSubmit={jest.fn()} />);
    
    expect(getByLabelText('Title:').value).toBe('Initial Title');
    expect(getByLabelText('Description:').value).toBe('Initial Description');
    expect(getByLabelText('Client URL:').value).toBe('https://www.example.com');
    expect(getByLabelText('Status:').value).toBe('visible');
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(<PortfolioForm initialData={initialData} onSubmit={jest.fn()} />);
    expect(asFragment()).toMatchSnapshot();
  });

});
