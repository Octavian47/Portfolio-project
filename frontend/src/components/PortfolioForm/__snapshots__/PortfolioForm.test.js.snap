// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PortfolioForm Component matches the snapshot 1`] = `
<DocumentFragment>
  <form
    class="portfolio-form"
  >
    <div>
      <label
        for="title"
      >
        Title:
      </label>
      <input
        id="title"
        required=""
        type="text"
        value="Initial Title"
      />
    </div>
    <div>
      <label
        for="description"
      >
        Description:
      </label>
      <textarea
        id="description"
        required=""
      >
        Initial Description
      </textarea>
    </div>
    <div>
      <label
        for="clientUrl"
      >
        Client URL:
      </label>
      <input
        id="clientUrl"
        placeholder="https://www.example.com"
        required=""
        type="url"
        value="https://www.example.com"
      />
      <small>
        Please include "https://" or "http://" in the URL.
      </small>
    </div>
    <div>
      <label
        for="status"
      >
        Status:
      </label>
      <select
        id="status"
        required=""
      >
        <option
          value="visible"
        >
          Display
        </option>
        <option
          value="hidden"
        >
          Hidden
        </option>
      </select>
    </div>
    <div>
      <label
        for="image"
      >
        Image:
      </label>
      <input
        id="image"
        type="file"
      />
    </div>
    <button
      type="submit"
    >
      Save
    </button>
  </form>
</DocumentFragment>
`;
